File: starterAIs/starter.ts
class Entity {
  TYPE_MONSTER = 0;
  TYPE_MY_HERO = 1;
  TYPE_OTHER_HERO = 2;
  MY_BASE = 1;
  OTHER_BASE = 2;
  distanceFromMyBase: number;
  constructor(
    public id: number,
    public type: number,
    public x: number,
    public y: number,
    public shieldLife: number,
    public isControlled: number,
    public health: number,
    public vx: number,
    public vy: number,
    public nearBase: number,
    public threatFor: number,
    private me: Player
  ) {
    this.distanceFromMyBase = this.getDistanceFrom(
      this.me.basePosX,
      this.me.basePosY
    );
  }
  isDangerousForMyBase = (): boolean => {
    return this.threatFor === this.MY_BASE;
  };
  getDistanceFrom = (x: number, y: number): number => {
    return Math.sqrt(Math.pow(x - this.x, 2) + Math.pow(y - this.y, 2));
  };
}

class Player {
  constructor(
    public basePosX: number,
    public basePosY: number,
    public baseHealth: number,
    public mana: number
  ) {}
  setHealth = (value: number) => {
    this.baseHealth = value;
  };
  setMana = (value: number) => {
    this.mana = value;
  };
  canCast = (): boolean => {
    return this.mana >= 10;
  };
}

class Game {
  ACTION_WAIT = "WAIT";
  ACTION_MOVE = "MOVE";
  ACTION_SPELL = "SPELL";
  SPELL_WIND = "WIND";
  SPELL_CONTROL = "CONTROL";
  SPELL_SHIELD = "SHIELD";

  me: Player;
  enemy: Player;
  entities: Entity[];

  constructor(baseX: number, baseY: number, private heroes: number) {
    this.me = new Player(baseX, baseY, 3, 0);
    this.enemy = new Player(
      baseX === 0 ? 17630 : 0,
      baseY === 0 ? 9000 : 0,
      3,
      0
    );
    this.entities = [];
  }

  newTurn = (
    health: number,
    mana: number,
    enemyHealth: number,
    enemyMana: number
  ) => {
    this.me.setHealth(health);
    this.me.setMana(mana);
    this.enemy.setHealth(enemyHealth);
    this.enemy.setMana(enemyMana);
    this.entities = [];
  };

  addEntity = (entity: Entity) => {
    this.entities.push(entity);
  };

  nextAction = (hero: number): string => {
    // In the first league: MOVE <x> <y> | WAIT; In later leagues: | SPELL <spellParams>;

    return this.ACTION_WAIT;
  };
  debug = (message: string, ...rest) => {
    console.error(message, ...rest);
  };
}

const [baseX, baseY] = readline().split(" ").map(Number); // The corner of the map representing your base
const heroesPerPlayer: number = Number(readline()); // Always 3
const game = new Game(baseX, baseY, heroesPerPlayer);

// game loop
while (true) {
  const myBaseInput: number[] = readline().split(" ").map(Number);
  const enemyBaseInput: number[] = readline().split(" ").map(Number);
  game.newTurn(
    myBaseInput[0],
    myBaseInput[1],
    enemyBaseInput[0],
    enemyBaseInput[1]
  );

  const entityCount: number = Number(readline()); // Amount of heros and monsters you can see
  for (let i = 0; i < entityCount; i++) {
    const inputs: number[] = readline().split(" ").map(Number);
    game.addEntity(
      new Entity(
        inputs[0], // Unique identifier
        inputs[1], // 0=monster, 1=your hero, 2=opponent hero
        inputs[2], // Position of this entity
        inputs[3],
        inputs[4], // Ignore for this league; Count down until shield spell fades
        inputs[5], // Ignore for this league; Equals 1 when this entity is under a control spell
        inputs[6], // Remaining health of this monster
        inputs[7], // Trajectory of this monster
        inputs[8],
        inputs[9], // 0=monster with no target yet, 1=monster targeting a base
        inputs[10], // Given this monster's trajectory, is it a threat to 1=your base, 2=your opponent's base, 0=neither
        game.me
      )
    );
  }
  for (let i = 0; i < heroesPerPlayer; i++) {
    console.log(game.nextAction(i));
  }
}

File: starterAIs/Starter.rb
class InputParser
    def getBaseLoc
        return gets.split(" ").collect { |coord| coord.to_i }
    end

    def getHeroesPerPlayer
        return gets.to_i
    end

    def getBaseDetails
        return gets.split(" ").collect { |detail| detail.to_i }
    end

    def getEntityCount
        return gets.to_i
    end

    def getEntityDetails
        return gets.split(" ").collect { |detail| detail.to_i }
    end
end

# ======================================================================================

class Base
    attr_accessor :x, :y, :health, :mana
    # @x        => x coord
    # @y        => y coord
    # @health   => base health
    # @mana     => Ignore in the first league; Spend ten mana to cast a spell

    def initialize(coords)
        @x, @y = coords
        @health = 0
        @mana = 0
    end

    def updateStats(stats)
        @health, @mana = stats
    end
end

# ======================================================================================

class Monster
    attr_accessor :id, :x, :y, :vx, :vy, :health, :locked_on, :threatening
    # @id           => identifier
    # @x            => x coord
    # @y            => y coord
    # @vx           => x trajectory
    # @vy           => y trajectory
    # @health       => remaining health
    # @locked_on    => 0=no target yet, 1=targeting a base
    # @threatening  => Given this monster's trajectory, it is a threat to 1=your base, 2=your opponent's base, 0=neither

    def initialize(stats)
        @id, @x, @y, @health, @vx, @vy, @locked_on, @threatening = stats
    end

    def is_threat?
        @threatening == 1
    end

end

# ======================================================================================

class Hero
    attr_accessor :id, :x, :y, :health
    # @id       => identifier
    # @x        => x coord
    # @y        => y coord
    # @health   =>  remaining health

    def initialize(stats)
        @id, @x, @y, @healthg = stats
    end

    def pickTarget(monsters)
        # replace below with your target picking logic here
        return monsters.first
    end
end

# ======================================================================================

class Game
    attr_accessor :input_parser, :base, :enemy_base, :monsters, :heroes

    def initialize()
        @inputParser = InputParser.new()
        @base = Base.new(@inputParser.getBaseLoc)
        @enemyBase = Base.new([17630, 9000]) # enemy base is in opposite corner in starting League
    end

    def playGame
        heroes_per_player = @inputParser.getHeroesPerPlayer

        # game loop
        loop do
            @monsters = [] # clear out the list of monsters from last turn
            @heroes = [] # clear out the list of heroes from last turn
            @base.updateStats(@inputParser.getBaseDetails)
            @enemyBase.updateStats(@inputParser.getBaseDetails)

            entity_count = @inputParser.getEntityCount # Amount of heros and monsters you can see
            entity_count.times do
                # id: Unique identifier
                # type: 0=monster, 1=your hero, 2=opponent hero
                # x: Position of this entity
                # shield_life: Ignore for this league; Count down until shield spell fades
                # is_controlled: Ignore for this league; Equals 1 when this entity is under a control spell
                # health: Remaining health of this monster
                # vx: Trajectory of this monster
                # near_base: 0=monster with no target yet, 1=monster targeting a base
                # threat_for: Given this monster's trajectory, is it a threat to 1=your base, 2=your opponent's base, 0=neither
                id, type, x, y, shield_life, is_controlled, health, vx, vy, locked_on, threatening = @inputParser.getEntityDetails
                
                if type == 0
                    @monsters << Monster.new([id, x, y, health, vx, vy, locked_on, threatening])
                end

                if type == 1
                    @heroes << Hero.new([id, x, y, health]) # may need to pass more in later leagues
                end

                if type == 2
                    # no op for now, we don't care about where enemy heroes are
                end

            end

            # To debug: STDERR.puts "Debug messages..."
            # Write an action using puts
            # In the first league: MOVE <x> <y> | WAIT; In later leagues: | SPELL <spellParams>;
            @heroes.each do |hero|
                if @monsters.size > 0
                    target = hero.pickTarget(@monsters)
                    puts "MOVE #{target.x} #{target.y} Attacking #{target.id} at (#{target.x},#{target.y})"
                else
                    puts "WAIT All clear, no monsters!"
                end
            end
        end
    end
end

STDOUT.sync = true # DO NOT REMOVE
game = Game.new()
game.playGame
